### ISSUE TITLE
Fatal Python error: GC object already tracked

### ISSUE THREAD

**Issue by drdangersimon** [blaze/dask#860]
Running python 2.7.10, dask 0.7.5 and pandas 0.17.1.

When I load a csv using dask.DataFrame.read_csv and try to do an apply on the array it gives me that error and stops python.

expample:
 import dask.dataframe as dd
chunks = dd.read_csv('path to csv', sep='\t')
data = chunks.apply(somefunction, axis=1, columns='somecolumn').compute()
Fatal Python error: GC object already tracked

**Comment by mrocklin** [comment]
I've seen similar issues when running Pandas 0.17.1 recently.  We actually have such an intermittent failure in our test suite now.  I suspect that recent GIL releasing within Pandas 0.17.1 is not entirely safe.

If you're able to provide a small example that always fails we should raise an issue on the Pandas issue tracker to see if they can resolve this issue.

**Comment by mrocklin** [comment]
cc @jreback @sinhrks

**Comment by mrocklin** [comment]
@drdangersimon you can avoid this issue if you operate with a single thread.  This can be accomplished globally with the following command

    import dask
    dask.set_options(get=dask.async.get_sync)

**Comment by drdangersimon** [comment]
Thanks for that fix it seems to not crash now. I've never encountered that bug with pandas only with dask, so i'm not sure how to find an example that always fails.

**Comment by jreback** [comment]
@mrocklin 

here is an exerciser for reading csv in parallel using pure pandas. pls try this out.
I am unable to reproduce the thread errors that a few people have seen.

```
import pandas as pd
from pandas.util import testing as tm
import numpy as np
from multiprocessing.pool import ThreadPool
pd.options.display.width=200
fn = 'test.csv'
np.random.seed(1234)

N = 1000000
df = pd.DataFrame(np.random.randn(N,5),columns=list('abcde'))
df['foo'] = 'foo'
df['bar'] = 'bar'
df['baz'] = 'baz'
df['date'] = pd.date_range('20000101 09:00:00',periods=N,freq='s')
df['int'] = np.arange(N)

df.to_csv(fn)

def f():
    return pd.read_csv(fn, index_col=0, header=0)

def g(procs=4):

    def reader(arg):
        start, nrows = arg

        # read from the start line for nrows
        if not start:
            return pd.read_csv(fn, index_col=0, header=0, nrows=nrows)

        # skip the header entirely
        return pd.read_csv(fn, index_col=0, header=None, skiprows=int(start)+1,nrows=nrows)

    # our tasks
    tasks = [ (N*i/procs, N/procs) for i in range(procs) ]
    pool = ThreadPool(processes=procs)

    # execute
    results = pool.map(reader, tasks)

    # collate the results
    # use the header from the first piece
    header = results[0].columns
    for r in results[1:]:
        r.columns = header
    return pd.concat(results)

def compare(result, expected):
    tm.assert_frame_equal(result, expected)
```

```
In [6]: run test.py

In [7]: expected = f()

In [8]: result = g(4)

In [9]: compare(result, expected)

In [10]: %timeit f()
1 loops, best of 3: 2.07 s per loop

In [11]: %timeit g(4)
1 loops, best of 3: 1.31 s per loop

In [12]: %timeit g(2)
1 loops, best of 3: 1.55 s per loop

In [13]: %timeit g(8)
1 loops, best of 3: 1.54 s per loop
```

**Comment by mrocklin** [comment]
@drdangersimon if you can provide a small csv file that fails with dask.dataframe I can try to translate it to a problem that fails under pandas (or I'll discover something wrong with how dask.dataframe is handling things.)

**Comment by drdangersimon** [comment]
Here is a script that throws the same error when I run it.
'''python
import numpy as np
import dask.dataframe as dd
# create larg file 1.8GB
f = open('largefile.csv','w')
for line in (i for i in np.random.rand(60*10**6)):
	f.write(str(line)+'\t'+str(line)+'\n')
f.close()
dask_frame = dd.read_csv('largefile.csv',sep='\t')
dask_frame[dask_frame.columns[0]].nunique().compute()
'''

**Comment by mrocklin** [comment]
I've raised issue https://github.com/pydata/pandas/issues/11786 on Pandas with a reproducible example that does not use `dask.dataframe`

**Comment by jreback** [comment]
PR to fix is here: https://github.com/pydata/pandas/pull/11790

pls give a test out (disclosure: not full fledged testing on pandas side atm).

**Comment by mrocklin** [comment]
Verified that this solves some `dask.dataframe.read_csv` problems.  

I'm also seeing excellent CPU utilization during this process.

```python
In [1]: import dask.dataframe as dd

In [2]: !du -hs largefile.csv
1.7G	largefile.csv

In [3]: df = dd.read_csv('largefile.csv', sep='\t', names=['a', 'b'])

In [4]: %time df.mean().compute()
CPU times: user 27.9 s, sys: 2.28 s, total: 30.2 s
Wall time: 4.93 s
Out[4]: 
a    0.49997
b    0.49997
dtype: float64

In [5]: 1.7e9 / 5 / 1e6  # MB/s
Out[5]: 340.0
```

**Comment by mrocklin** [comment]
That time goes up a bit with object dtype present (which presumably locks the GIL)

```python
In [7]: df = dd.read_csv('largefile.csv', sep='\t', names=['a', 'b'], dtype={'a': float, 'b': object})

In [8]: %time df.a.mean().compute()
CPU times: user 44.8 s, sys: 3.78 s, total: 48.6 s
Wall time: 30.2 s
Out[8]: 0.4999699105488524

In [9]: df = dd.read_csv('largefile.csv', sep='\t', names=['a', 'b'], dtype={'a': float, 'b': object}, usecols=['a'])

In [10]: %time df.a.mean().compute()
NCPU times: user 20 s, sys: 2.11 s, total: 22.1 s
Wall time: 3.67 s
Out[10]: 0.4999699105488524
```

**Comment by jcrist** [comment]
With https://github.com/pydata/pandas/pull/11790 merged, can this be closed?

**Comment by mrocklin** [comment]
Pandas will maybe released in two weeks.  Sure, lets close.  I don't imagine we'll do anything else before then.
